docs/Style.thy

```diff
@@ -0,0 +1,260 @@
+      compatible, diffs are much easier to read. Pull requests mean we read a lot of diffs.
+      grep and other regexp search is not needed as much as it used to, but it's still often
+      useful to be able to find constants and lemmas by regexp search outside of Isabelle
+
+  * Don't fight existing auto-indent where possible:
+      there can be exceptions, but each time we go against defaults, it will be annoying to follow.
+
+  * Compatibility with upstream style where it exists:
+      there is not much consistent Isabelle distribution style, but where there is, we should follow
+      it, so that others can read our code and we can contribute upstream
+\<close>
+
+text \<open>
+The main rules:
+  * Indent your proofs.
+  * Follow naming conventions.
+  * Be aware that others have to read what you write.
+
+To not drive proof maintainers insane:
+  * Do not use auto except at the end of a proof.
+  * Never check in proofs containing back.
+  * Instantiate bound variables in preference to leaving schematics in a subgoal
+    (i.e. use erule_tac ... allE in preference to erule allE)
+  * Explicitly name variables that the proof text refers to (e.g. with rename_tac)
+  * Don't mix object and meta logic in a lemma statement.
+\<close>
+
+section \<open>Indentation\<close>
+
+text \<open>
+Isabelle code is much easier to maintain when indented consistently. In apply style proofs we
+indent by 2 spaces, and add an additional space for every additional subgoal. For example, the
+rules iffI and conjI add a new subgoal, and fast removes a subgoal. The idea is that, when
+something breaks, the indentation tells you whether a tactic used to solve a subgoal or produced
+new ones.
+\<close>
+
+lemma disj_conj_distribR:
+  "(A \<and> B \<or> C) = ((A \<or> C) \<and> (B \<or> C))"
+  apply (rule iffI)
+   apply (rule conjI)
+    apply fast
+   apply fast
+  apply auto
+  done
+
+text \<open>
+The statement of Hoare triple lemmas are generally indented in the following way.
+\<close>
+
+lemma my_hoare_triple_lemma:
+  "\<lbrace> precondition_one and precondition_two and
+     precondition three \<rbrace>
+   my_function param_a param_b
+   \<lbrace> post_condition \<rbrace>"
+  oops
+
+text \<open> or \<close>
+
+lemma my_hoare_triple_lemma:
+  "\<lbrace> precondition_one and precondition_two
+     and precondition three \<rbrace>
+   my_function param_a param_b
+   \<lbrace> post_condition \<rbrace>"
+  oops
+
+text \<open>
+Definitions, record and datatypes are generally indented as follows and use \<equiv> rather than =. Short
+datatypes can be on a single line.
+\<close>
+
+definition word_bits :: nat where
+  "word_bits \<equiv> 32"
+
+definition long_defn_type ::
+  "'this => 'is => 'a => 'long => 'type => 'that => 'does =>
+   'not => 'fit => 'in => 'one => 'line"
+  where
+  "long_defn_type \<equiv> undefined"
+
+fun foo_fun :: "nat \<Rightarrow> nat" where
+  "foo_fun 0 = 0" |
+  "foo_fun (Suc n) = n"
```

Comment:
I know that this came from the previous document, but my strong preference is for
```suggestion
fun foo_fun :: "nat \<Rightarrow> nat" where
    "foo_fun 0 = 0"
  | "foo_fun (Suc n) = n"
```

Reply:
I used to write them like this and still have. a slight preference for it, but the other style seems more popular upstream. I'm easy on this one. If people prefer it, happy with that.